Increase packages/components test coverage

nikomakela opened the pull request:
TH-1409

nikomakela committed:
chore(sonar): exclude packages/components files from test coverage

TH-1409

nikomakela committed:
test: apollo components

TH-1409

nikomakela committed:
test: base app

TH-1409

nikomakela committed:
test: useCmsHelper hook

TH-1409

nikomakela committed:
test: useCmsRoutedAppHelper hook

TH-1409

nikomakela committed:
test: useApolloErrorHandler

TH-1409

nikomakela committed:
chore(sonar): resync test coverage exclusions to sonar configuration

TH-1409

nikomakela committed:
test: askemFeedbackContainer

TH-1409

nikomakela committed:
test: useAskem

TH-1409

nikomakela committed:
refactor: optimize useAskemContext rerendering

TH-1409.

The unit tests proved that the context hook gets rendered 3 times when
it should render only twice.

Added unit tests for the hook.

nikomakela committed:
test: basicMeta

TH-1409

nikomakela committed:
fix: meta tags for graphs, twitter and canonical urls

TH-1409.

1. The twitter meta tags properties should be set in name-attribute.
2. The OpenGraph meta tag properties should be set in
property-attribute.
3. Canonical URL is now created with a hook that has improved handling
for origin.
4. Alternate URLs might have had double locales.

nikomakela committed:
test: geolocationService

TH-1409.

nikomakela committed:
test: formatDateTimeIntoLocaleString

TH-1409

nikomakela committed:
test: getDateArray

TH-1409

nikomakela committed:
test: getDomain

TH-1409

nikomakela committed:
test: getEnvOrError

TH-1409

nikomakela committed:
refactor: improve getSecureImage features

TH-1409.

Original `getSecureImage` has been copied from the original
events-helsinki-ui repository.
See https://github.com/City-of-Helsinki/events-helsinki-ui/
blame/18eabe18acd16fe84d1959a019ccb12b6416dd8b/src/util/getSecureImage.tsx

NOTE: the `getSecureImage` is using https://images.weserv.nl/ as an
image proxy, that is not validated by the organisation. However, it's
open sourced and BSD 3 licensed and has been working well for years now,
so it's left as an option. It might work well as an cache, to reduce
queries made to LinkedEvents or other services where the images are
hosted. All the event iamges are public anyway.

nikomakela committed:
test: headless-cms service

TH-1409

nikomakela committed:
test: list component

TH-1409

terovirtanen commented:
**TAPAHTUMAT-UI branch is deployed to platta: https://tapahtumat-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**HARRASTUKSET-UI branch is deployed to platta: https://harrastukset-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**VENUE-GRAPHQL-PROXY branch is deployed to platta: https://venue-graphql-proxy-pr823.dev.hel.ninja** :rocket::rocket::rocket:

karisal-anders committed:
test(components): fix types in List.test.tsx

Before this change "yarn typecheck" failed in packages/components with:
```
src/components/list/__tests__/List.test.tsx:107:54 - error TS2322: Ty...
  Type '({ children, className, }: ListItemProps & { children: React....
    Types of parameters '__0' and 'props' are incompatible.
      Type 'ListItemProps' is not assignable to type 'ListItemProps &...
        Property 'children' is missing in type 'ListItemProps' but
        required in type '{ children: ReactNode; }'.

107  const { container } = render(<List items={items} li={CustomLi} />);
                                                         ~~
  src/components/list/__tests__/List.test.tsx:104:26
    104     }: ListItemProps & { children: React.ReactNode }) => (
                                 ~~~~~~~~
    'children' is declared here.
  src/components/list/List.tsx:22:3
    22   li?: React.ComponentType<ListItemProps>;
         ~~
    The expected type comes from property 'li' which is declared here on
    type 'IntrinsicAttributes & Props'
```

refs TH-1409

karisal-anders committed:
test(components): fix imports & types in parseRedirects.test.tsx

Before this change "yarn typecheck" failed in packages/components with:
```
src/edge-runtime-compatible/middlewares/shortUrls/dynamicUrls/__tests__/
parseRedirects.test.tsx:2:28 - error TS2305: Module
'"../../../../types"' has no exported member 'Redirects'.

2 import type { AppLanguage, Redirects } from '../../../../types';
                             ~~~~~~~~~
```

refs TH-1409

karisal-anders committed:
fix(vitest): replace use of deprecated cache.dir with cacheDir

Running tests before this change caused a deprecation message to appear:
```
DEPRECATED  "cache.dir" is deprecated, use Vite's "cacheDir" instead if
you want to change the cache director. Note caches will be written to
"cacheDir/vitest"
```

Done:
- Replace cache.dir with cacheDir in vitest configurations
- Remove the `/vitest` from the paths as the deprecation message says:
  `Note caches will be written to "cacheDir/vitest"`
  - Tested by running "yarn test" under packages/components and
    packages/graphql-proxy-server, and checked that at the repository's
    root .cache/events-helsinki-components/vitest and
    .cache/graphql-proxy-server/vitest appeared
- Fix `graphql-proxy-server` component's cacheDir (It looked incorrect)

From https://vitest.dev/config/#cache documentation:
```
The cache directory is controlled by the Vite's cacheDir option:

import { defineConfig } from 'vitest/config'

export default defineConfig({
  cacheDir: 'custom-folder/.vitest'
})

You can limit the directory only for Vitest by using process.env.VITEST:

import { defineConfig } from 'vitest/config'

export default defineConfig({
  cacheDir: process.env.VITEST ? 'custom-folder/.vitest' : undefined
})
```

refs TH-1409

karisal-anders committed:
test(components): fix service.test.ts by fixing listGeneratedPages paths

Before this change running "yarn test" under packages/components on
Windows caused:
```
 FAIL  src/utils/headless-cms/__tests__/service.test.ts >
 headless-cms/service > NextPageRevalidateService > NextPageLister
 functionality > listGeneratedPages should list, deduplicate, and sort
 files recursively
AssertionError: expected [ '/fi\about', '/fi\courses\course1' ] to
strictly equal [ '/fi/about', '/fi/courses/course1' ]

- Expected
+ Received

  [
-   "/fi/about",
-   "/fi/courses/course1",
+   "/fi\\about",
+   "/fi\\courses\\course1",
  ]

 ❯ src/utils/headless-cms/__tests__/service.test.ts:326:23
    324|           path.join(mockProjectRoot, '.next', 'server', 'pag...
    325|         );
    326|         expect(pages).toStrictEqual(
       |                       ^
    327|           ['/fi/about', '/fi/courses/course1'].sort()
    328|         );
```

So the path handling was platform dependent, even though the result was
wanted to be always using forward slashes as path separator. Now
listGeneratedPages converts backslashes to forward slashes, so the
result is the same on Windows too.

refs TH-1409 (path handling was platform dependent, now it's not)

karisal-anders committed:
chore(components): fix linting errors and warnings

Before this change "yarn lint" under packages/components caused:
```
src/cmsHelperProvider/__tests__/useCmsRoutedAppHelper.test.tsx
  18:5  warning  'path' is defined but never used. Allowed unused args
  must match /^_/u  @typescript-eslint/no-unused-vars

src/components/domain/event/similarEvents/__tests__/SimilarEvents.tes...
  13:1  error  `jest-date-mock` import should occur before import of
  `@/test-utils`  import/order
  14:1  error  `react` import should occur before import of
  `@/test-utils`           import/order

src/utils/headless-cms/__tests__/service.test.ts
  439:15  warning  'listGeneratedPagesSpy' is assigned a value but never
  used. Allowed unused vars must match /^_/u
  @typescript-eslint/no-unused-vars

✖ 4 problems (2 errors, 2 warnings)
```

Done:
- Reordered the imports
- Added underscore in front of the unused parameter/variable to get rid
  of the unused variable linting warning

refs TH-1409

karisal-anders committed:
chore(components): remove unneeded @ts-expect-error and @ts-ignore

fix some of the types in BaseApp.test.tsx and useAskemContext.test.tsx
soe @ts-expect-error or @ts-ignore are not needed

refs TH-1409

nikomakela committed:
docs: add README for packages/components

TH-1409 TH-1422 TH-1340.

nikomakela committed:
docs: add README for packages/graphql-proxy-server

TH-1409 TH-1422.

karisal-anders commented on packages/components/src/apollo/__tests__/EventsFederationApolloClient.test.ts:
```diff
@@ -0,0 +1,66 @@
+    });
+  });
+
+  it('should allow unauthorized requests if configured', () => {
+    const config = {
+      ...baseConfig,
+      allowUnauthorizedRequests: true,
+      federationGraphqlEndpoint: 'https://example.com/graphql',
+    };
+    const client = new EventsFederationApolloClient(config);
+    const httpLink = client.getHttpLink(config.federationGraphqlEndpoint);
+    expect(httpLink).toBeDefined();
+  });
```
getHttpLink function is:
```typescript
  getHttpLink(uri: string) {
    let options: HttpOptions = {
      uri,
      fetch,
    };

    // Our review environment can't provide a valid certificate. Hence we allow
    // the application to be configured so that unauthorized requests are not
    // rejected. This allows us to test API changes in the review environment.
    if (
      this.config.allowUnauthorizedRequests &&
      new URL(uri).protocol === 'https:'
    ) {
      graphqlClientLogger.info('Allowing unauthorized requests');
      options = {
        ...options,
        fetchOptions: {
          agent: new https.Agent({ rejectUnauthorized: false }),
        },
      };
    }

    return new HttpLink(options);
  }
```

Is there some validation in HttpLink constructor that could make the return value undefined?

nikomakela replied:
It should probably test the httpLink's fetchOptions instead.

terovirtanen commented:
**FEDERATION-ROUTER branch is deployed to platta: https://events-graphql-federation-sports-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**FEDERATION-ROUTER branch is deployed to platta: https://events-graphql-federation-events-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**HARRASTUKSET-UI branch is deployed to platta: https://harrastukset-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**TAPAHTUMAT-UI branch is deployed to platta: https://tapahtumat-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**VENUE-GRAPHQL-PROXY branch is deployed to platta: https://venue-graphql-proxy-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**EVENTS-GRAPHQL-PROXY branch is deployed to platta: https://events-graphql-proxy-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr823.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr823.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __failed__ for https://tapahtumat-pr823.dev.hel.ninja :crying_cat_face::anger::boom::boom:

nikomakela committed:
ci(components): fix code quality workspace commands

TH-1410.

nikomakela committed:
test: fix eventsFederationApolloClient rejectUnauthorized test

TH-1409

nikomakela committed:
style: fix linting issues in import order

TH-1409

karisal-anders commented on packages/components/src/components/domain/seo/meta/__tests__/RouteMeta.test.tsx:
```diff
@@ -0,0 +1,183 @@
+  });
+
+  describe('with page prop', () => {
+    const mockPage = {
+      id: '1',
+      title: 'Test Page',
+      slug: 'test-page',
+      uri: '/fi/test-page/',
+      link: '/fi/test-page/',
+      contentTypeName: 'Page',
+      date: '2023-01-01',
+      modified: '2023-01-01',
+      language: {
+        code: 'FI',
+        id: '1',
+        locale: 'fi',
+        name: 'Suomi',
+        slug: 'fi',
+      },
+      translations: [
+        {
+          uri: '/sv/test-sida/',
+          language: { slug: 'sv' },
+        },
+        {
+          uri: '/en/test-page/',
+          language: { slug: 'en' },
+        },
+      ],
+    } as unknown as PageType;
```
Removing unneeded fields (date, modified) and adding language.id fields, using enum for code makes the types for mockPage pass:
```suggestion
    const mockPage = {
      id: '1',
      title: 'Test Page',
      slug: 'test-page',
      uri: '/fi/test-page/',
      link: '/fi/test-page/',
      __typename: 'Page',
      language: {
        code: LanguageCodeEnum.Fi,
        id: '1',
        locale: 'fi',
        name: 'Suomi',
        slug: 'fi',
      },
      translations: [
        {
          uri: '/sv/test-sida/',
          language: { id: '1', slug: 'sv' },
        },
        {
          uri: '/en/test-page/',
          language: { id: '2', slug: 'en' },
        },
      ],
    } as const satisfies PageType;
```

terovirtanen commented:
**FEDERATION-ROUTER branch is deployed to platta: https://events-graphql-federation-hobbies-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**FEDERATION-ROUTER branch is deployed to platta: https://events-graphql-federation-hobbies-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**FEDERATION-ROUTER branch is deployed to platta: https://events-graphql-federation-sports-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**FEDERATION-ROUTER branch is deployed to platta: https://events-graphql-federation-events-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**TAPAHTUMAT-UI branch is deployed to platta: https://tapahtumat-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**EVENTS-GRAPHQL-PROXY branch is deployed to platta: https://events-graphql-proxy-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**LIIKUNTA-UI branch is deployed to platta: https://liikunta-pr823.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr823.dev.hel.ninja :laughing::tada::tada::tada:

nikomakela commented on packages/components/src/geolocation/__tests__/GeolocationService.test.ts:
```diff
@@ -0,0 +1,80 @@
+  };
+
+  beforeEach(() => {
+    vi.stubGlobal('navigator', {
+      geolocation: mockGeolocation,
+    });
+  });
+
+  afterEach(() => {
+    vi.restoreAllMocks();
```
> You can set global variable by assigning a value to globalThis or using [vi.stubGlobal](https://vitest.dev/api/vi#vi-stubglobal) helper. When using vi.stubGlobal, it will not automatically reset between different tests, unless you enable [unstubGlobals](https://vitest.dev/config/#unstubglobals) config option or call [vi.unstubAllGlobals](https://vitest.dev/api/vi#vi-unstuballglobals).
https://vitest.dev/guide/mocking.html#mock-a-global-variable

Yes, it seems we need to.

karisal-anders commented on packages/components/sonar-project.properties:
```diff
@@ -9,23 +9,37 @@ sonar.exclusions = \
   src/styles/**, \
   src/types/**, \
+  src/loggers/**, \
+  src/edge-runtime-compatible/**, \
```
What is the argument for excluding these? They have tests already, although not 100% coverage.

karisal-anders commented on packages/components/vitest.config.ts:
```diff
@@ -54,6 +52,15 @@ export default defineConfig(({ mode }) => ({
         'src/styles/**',
         'src/types/**',
+        '**/*.query.ts',
+        'src/edge-runtime-compatible/**',
```
Same as in https://github.com/City-of-Helsinki/events-helsinki-monorepo/pull/823/files#r2336917246 i.e. why should this folder be excluded?

karisal-anders reviewed (approved):
Read (and partly skimmed) through the whole thing—whew! Left some comments, please take a look at them.

nikomakela committed:
test: unstub global variables

TH-1409.

nikomakela committed:
test: include edge-runtime-compatible directory in test coverage

TH-1409
